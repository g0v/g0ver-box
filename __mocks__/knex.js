import knex from 'jest-mock-knex';

export { client } from 'jest-mock-knex';

export default knex;

process.setMaxListeners(0);
