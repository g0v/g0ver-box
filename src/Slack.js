import _ from 'lodash';
import slack from 'slack';

export const bot = slack.rtm.client();

const TOKEN = process.env.SLACK_TOKEN;

if (TOKEN) {
  bot.listen({ token: TOKEN });
}

bot.hello((message) => {
  console.log(`Got a message: ${message.type}`);
});

const call = (name, req = {}) => (
  new Promise((resolve, reject) => {
    const fn = _.get(slack, name);
    fn(
      { ...req, token: TOKEN },
      (err, res) => (err ? reject(err) : resolve(res))
    );
  })
);

export default {
  userInfo: async req => await call('users.info', req),
  postMessage: async req => await call('chat.postMessage', { ...req, as_user: true }),
  postMultiMessage: async (users, message) => _.map(users, user => bot.ws.send(JSON.stringify({
    id: `${user}::${Date.now()}`, type: 'message', channel: user, text: message,
  }))),
  channelInfo: async req => await call('channels.info', req),
  channelJoin: async req => await call('channels.join', req),
};
