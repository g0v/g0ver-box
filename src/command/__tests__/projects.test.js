/* eslint import/imports-first: 0 */

jest.mock('../../../knexfile', () => ({
  client: 'pg',
  connection: {
    host: '127.0.0.1',
    user: 'postgres',
    password: null,
    database: 'g0ver_projects',
  },
}));

import { client } from 'knex';
import Slack from '../../Slack';
import db from '../../model/Database';
import index from '../';

describe('projects command', () => {
  const data = { channel: 'D100', user: 'U03B2AB13', name: 'yutin', text: 'projects' };

  it('successful query', async () => {
    client.mockReturnValueOnce([{
      id: 10,
      title: 'g0ver box',
      userId: 'U03B2AB13',
      archive: {
        tags: '媒合 挖坑 推坑',
        url: 'https://g0v.tw/',
        thumb: 'https://g0v.tw/thumb.gif',
      },
    }, {
      id: 15,
      title: 'g0v today',
      userId: 'U03B2AB00',
      archive: {
        url: 'https://g0v.tw/',
      },
    }]);

    await index(data);
    expect(client).toMatchSnapshot();
    expect(client).toHaveBeenCalledTimes(1);
    expect(Slack.postMessage.mock.calls).toMatchSnapshot();
  });

  describe('database', () => {
    beforeAll(async () => {
      await db('pg_tables').select('tablename').where('schemaname', 'public').map(({ tablename }) => (
        db.schema.dropTable(tablename)
      ));
      await db.migrate.latest();
    });

    afterAll(async () => {
      await db.destroy();
    });

    it('create project', async () => {
      await index({ ...data, text: 'create g0ver-box' });
      await index({ ...data, text: 'https://project.g0v.tw/' });
      await index({ ...data, text: '找人 找坑' });
      await index({ ...data, text: 'https://project.g0v.tw/index.png' });
      await index({ ...data, text: 'yes' });
      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100',
        text: 'Done, create g0ver-box',
      });
    });

    it('projects', async () => {
      await index(data);
      expect(Slack.postMessage.mock.calls).toMatchSnapshot();
    });

    it('remove projects', async () => {
      await index({ ...data, text: 'remove g0ver-box' });
      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100',
        text: 'Done, 刪除 g0ver-box 專案',
      });
    });
  });
});
