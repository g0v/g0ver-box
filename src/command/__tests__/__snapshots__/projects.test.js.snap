// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projects command database projects 1`] = `
Array [
  Array [
    Object {
      "attachments": Array [
        Object {
          "color": "#000",
          "mrkdwn_in": Array [
            "text",
            "pretext",
            "fields",
          ],
          "text": "坑主: <@U03B2AB13>
找人 找坑",
          "thumb_url": "https://project.g0v.tw/index.png",
          "title": "g0ver-box",
          "title_link": "https://project.g0v.tw/",
        },
      ],
      "channel": "D100",
      "text": "下列專案歡迎你的加入：",
    },
  ],
]
`;

exports[`projects command successful query 1`] = `
Array [
  "select * from project where deleted_at is null limit 1000",
]
`;

exports[`projects command successful query 2`] = `
Array [
  Array [
    Object {
      "attachments": Array [
        Object {
          "color": "#000",
          "mrkdwn_in": Array [
            "text",
            "pretext",
            "fields",
          ],
          "text": "坑主: <@U03B2AB13>
媒合 挖坑 推坑",
          "thumb_url": "https://g0v.tw/thumb.gif",
          "title": "g0ver box",
          "title_link": "https://g0v.tw/",
        },
        Object {
          "color": "#000",
          "mrkdwn_in": Array [
            "text",
            "pretext",
            "fields",
          ],
          "text": "坑主: <@U03B2AB00>",
          "thumb_url": null,
          "title": "g0v today",
          "title_link": "https://g0v.tw/",
        },
      ],
      "channel": "D100",
      "text": "下列專案歡迎你的加入：",
    },
  ],
]
`;
